Add sum-all program

vaibhavarora14 opened the pull request:
**Program strategy -**
- Sorts input array to find `smallNumber` at `0` position and `largerNumber` at `1` position.
- Get a range generator object to fetch numbers between `small` and `large` number of input array.
- Iterate over generator and `add` them to a variable, which is meant to start from 0 and increase as numbers getting added.
- Return value which was holding result from `addition` operation in previous step.

Commit:
#3 Add README for sum-all program

Commit:
Add relevant test for sum-all method 😎

Commit:
\#3: WIP sum-all implementation with
- keeping only positive test in focus
- and some expected exceptions to throw

Commit:
Fix problem of getting +1 incremented value
- it was getting added in lower range in 1st step, resulting in wrong results of program

Commit:
\#3: Handle decimal numbers by
- flooring them to integers, as program counts number between integers excluding decimals coming in between

ArfatSalman commented on js-exercises/sum-all/sumAll.js:
```diff
@@ -0,0 +1,44 @@
+const isNumber = (arg) => typeof arg === 'number';
+
+const areNumbers = (args) => args.every(isNumber);
```
This would be true for `NaN` as well.

vaibhavarora14 replied:
right, issue fixed in 1e9a59a7b31f1961f18cfa90dab8749e73646be6

ArfatSalman commented on js-exercises/sum-all/sumAll.js:
```diff
@@ -0,0 +1,44 @@
+
+const areNumbers = (args) => args.every(isNumber);
+
+function* getRange(start, end) {
```
A for loop is preferred since it is less likely to make an infinite loop there.

ArfatSalman replied:
Also, start can have a good default value such as 0.
```suggestion
function* getRange(start = 0, end) {
```

ArfatSalman commented on js-exercises/sum-all/sumAll.js:
```diff
@@ -0,0 +1,44 @@
+
+function* getRange(start, end) {
+  let temp = start;
+  while (temp <= end) {
```
It is really important to consider whether whether you want to include the last element as you did here, to not include it. For the record, python's range does not include the last element.

vaibhavarora14 replied:
yes, right.
current scope of sum-all specs says to return sum of all numbers between `start`, `end`, including them.`getRange` currently works according to that.

If `getRange` has to grow further. it should be entertaining both types of requirement.
- return numbers between `start` and `end`. (Note: Excluding `start` as well unlike Python one).
- return numbers between `start` and `end`, including parameters passed.

What say?

ArfatSalman commented on js-exercises/sum-all/sumAll.js:
```diff
@@ -0,0 +1,44 @@
+}
+
+function sumAll(arrayOfNumbers) {
+  let inputArray = arrayOfNumbers;
```
Why alias `arrayOfNumbers`?

vaibhavarora14 replied:
variable names improved with b093b7825633e962b20ad8258369b28eaa7ee122

ArfatSalman commented on js-exercises/sum-all/sumAll.js:
```diff
@@ -0,0 +1,44 @@
+  const sortedArray = inputArray.sort((a, b) => a - b);
+
+  let result = 0;
+  const rangeIterator = getRange(sortedArray[0], sortedArray[1]);
```
```suggestion
  const rangeIterator = getRange(lowerBound, upperBound);
```

vaibhavarora14 replied:
You were looking for something like
21c5e9e924a14adc1d49eca299ae56131e5a938e && a984294784b6516bdf08e2ee64facfecc46069ab ?

ArfatSalman commented on js-exercises/sum-all/sumAll.js:
```diff
@@ -0,0 +1,44 @@
+  let result = 0;
+  const rangeIterator = getRange(sortedArray[0], sortedArray[1]);
+
+  for (const number of rangeIterator) {
```
```suggestion
  for (const number of getRange(lowerBound, upperBound)) {
```

ArfatSalman reviewed (changes requested):
Also try the arithmetic mean formula.

Commit:
#3 Fix issue with isNumber() for isNaN argument

vaibhavarora14 committed:
Set default value in case, nothing comes in `start`

Co-Authored-By: Arfat Salman <giney.paradise@gmail.com>

Commit:
#3 Improve variable names

Commit:
remove unnecessary variable declaration

Commit:
Merge branch 'sum-all' of https://github.com/pesto-students/deep-dive-varora1406 into sum-all

Commit:
#3 Make code more readable with lower && upper bound variables

Commit:
#3 Make code more loosly coupled
- Extract out "addition" operation out from sub-all method
- Remove for loop && used spread operation with generator function, to reduce any chances of mistakes.

Commit:
#3 Add default value for end as well of getRange

vaibhavarora14 commented:
@ArfatSalman I didn't understood what you want to say with https://github.com/pesto-students/deep-dive-varora1406/pull/4#pullrequestreview-400413577.

Should I make program for mean, as sum-all?
OR did you wanted me to realize about tight coupled code here, which i recently loosen up with https://github.com/pesto-students/deep-dive-varora1406/pull/4/commits/73146eb2376b8b8ccaea6604211f76fcd31ef12b
